### PR TITLE
view/xwayland: avoid focusing views that don't want focus

### PR DESCRIPTION
**Pull request by jlindgren90**
XWayland views can self-declare that they don't want keyboard focus via
the ICCCM WM_HINTS property. Most of the logic is already in place to
avoid giving focus to such views (e.g. taskbars).

Add a couple of missing pieces to make this work:

- Hook up `view_isfocusable()` to look at WM_HINTS for XWayland views
- Adjust `desktop_focus_topmost_mapped_view()` to skip unfocusable views

I also noticed that in several places, we will still focus and
raise a view regardless of `view_isfocusable()`. To fix this, make
`desktop_focus_and_activate_view()` return false if the view isn't
focusable, and don't raise the view to the front in that case.
    
Possible FIXME: a couple of places still call `workspaces_switch_to()`
before `desktop_focus_and_activate_view()`. Maybe switching workspaces
should also be prevented if the view isn't focusable?

### CODE REVIEW

**Comment by Consolatis** [comment]
Didn't test but code and idea looks good to me.
On a related note: there were also requests to add a window rule property which prevents keyboard focus completely, this PR looks like it would make it pretty easy to implement.

> Possible FIXME: a couple of places still call `workspaces_switch_to()`
before `desktop_focus_and_activate_view()`. Maybe switching workspaces
should also be prevented if the view isn't focusable?

Yes, absolutely.
Maybe we could refactor that a bit to just do the workspace check and switch in `desktop_focus_and_activate_view()` directly rather than having multiple places which call `desktop_focus_and_activate_view()` do it themselves.

**Comment by jlindgren90** [comment]
> Yes, absolutely. Maybe we could refactor that a bit to just do the workspace check and switch in `desktop_focus_and_activate_view()` directly rather than having multiple places which call `desktop_focus_and_activate_view()` do it themselves.

Should `desktop_focus_and_activate_view()` always switch to the workspace of the view being activated? Right now it's inconsistent, e.g. `xdg_activation_handle_request()` does switch but xwayland's `handle_request_activate()` doesn't.

I never got used to using more than one workspace myself, so I'm not sure what's the expected behavior here.

**Comment by Consolatis** [comment]
I can't imagine a situation where one would want to have keyboard focus on a window that is on a different workspace and thus not visible. Activation via foreign toplevel protocol (e.g. click on a panel) also switches workspaces manually currently.

**Comment by jlindgren90** [comment]
Makes sense. In some cases (like un-minimizing) I could imagine wanting the activated window moved to the current workspace, rather than switching workspaces. But labwc doesn't have that option currently, so probably a moot point.

**Comment by jlindgren90** [comment]
Marking this as draft for now. There's another focus issue that ought to be fixed first (pull request coming soon).

**Comment by jlindgren90** [comment]
v2:
- Rebase on top of #1103
- Rename `desktop_focus_and_activate_view()` to simply `desktop_focus_view()`.
- Move switching workspaces and (optionally) raising the view inside `desktop_focus_view()`, so these steps are also skipped for unfocusable views.
- Move the "unfocus if no focusable view" logic from `desktop_focus_view()` to `desktop_focus_topmost_mapped_view()` (now `desktop_focus_topmost_view()`) as we don't want clicking on an unfocusable view (like a taskbar) to unfocus the current view, but do want it unfocused if we switch to an empty workspace.

**Comment by jlindgren90** [comment]
v3: Rebased on top of #1109

**Comment by jlindgren90** [comment]
Should be ready for review now.

**Comment by Consolatis** [comment]
Code LGTM, I'd like to give it a test ride tonight though to ensure we don't regress on workspace related stuff.

File: src/xwayland.c [hunk 1]
```diff
@@ -30,6 +30,16 @@ xwayland_view_get_size_hints(struct view *view)
...
 }
 
+static bool
+xwayland_view_wants_focus(struct view *view)
+{
+	xcb_icccm_wm_hints_t *hints = xwayland_surface_from_view(view)->hints;
+	if (!hints) {
+		return true;
+	}
+	return (bool)hints->input;
```
Comment (johanmalm) [inline src/xwayland.c:40]: How confident are we on this logic (`hints->input`)?
Having a quick read here (https://tronche.com/gui/x/icccm/sec-4.html) makes my head spin.

**Comment by johanmalm** [comment]
This was a new one one me (xwayland views being able to not want keyboard focus)
Is there a particular use-case for this where the surface isn't an override-redirect (OR) surface?

For reference (anyone else reading/reviewing) - this is how wlroots deals with OR surfaces.
https://gitlab.freedesktop.org/wlroots/wlroots/-/blob/master/xwayland/xwm.c?ref_type=heads#L2284

File: include/workspaces.h [hunk 2]
```diff
@@ -2,6 +2,9 @@
 #ifndef LABWC_WORKSPACES_H
 #define LABWC_WORKSPACES_H
 
+#include <stdbool.h>
+#include <wayland-util.h>
```
Comment (johanmalm) [inline include/workspaces.h:6]: Why `wayland-util.h` here? 
The `struct wl_list` is already forward declared.

**Comment by johanmalm** [comment]
That second commit is hard to follow (for me reading this late in the evening anyway :smile:)
Is it worth splitting it out into smaller commits doing fewer things?

File: src/xwayland.c (hunk 1 above)
Comment (jlindgren90) [reply src/xwayland.c:40]: Speaking for myself, no more or less confident than anything else X11... it's all trial-and-error.
This is how `xprop` describes that flag, for what it's worth:
```
WM_HINTS(WM_HINTS):
		Client accepts input or input focus: False
```

Comment (jlindgren90) [reply src/xwayland.c:40]: There is some more complexity with a WM_TAKE_FOCUS atom that I don't fully understand either; however, I think we are using the "input" flag (which is just a Boolean, though 32 bits) correctly so far as it goes:

> Passive and Locally Active clients set the input field of WM_HINTS to True, which indicates that they require window manager assistance in acquiring the input focus. No Input and Globally Active clients set the input field to False, which requests that the window manager not set the input focus to their top-level window.

Comment (johanmalm) [reply src/xwayland.c:40]: Happy to try it :smile:
Most of the code I've written for labwc has been trial-and-error tbh. Certain things are clear in protocols, etc but there is a lot of things relating to window management that aren't exactly well documented nor obvious.

**Comment by jlindgren90** [comment]
> This was a new one one me (xwayland views being able to not want keyboard focus) Is there a particular use-case for this where the surface isn't an override-redirect (OR) surface?

Some X11 panels, for example `lxqt-panel` or my fork of it.
Oddly, `xfce4-panel` sets `WM_HINTS.input = true` but still does not get keyboard focus under Openbox (not sure why).

File: include/workspaces.h (hunk 2 above)
Comment (jlindgren90) [reply include/workspaces.h:6]: Forward declaration is not enough for the `struct wl_list link;` declaration below; the size of the struct needs to be known.

File: src/xwayland.c (hunk 1 above)
Comment (johanmalm) [reply src/xwayland.c:40]: > There is some more complexity with a WM_TAKE_FOCUS atom that I don't fully understand either; however, I think we are using the "input" flag (which is just a Boolean, though 32 bits) correctly so far as it goes:
> 
> > Passive and Locally Active clients set the input field of WM_HINTS to True, which indicates that they require window manager assistance in acquiring the input focus. No Input and Globally Active clients set the input field to False, which requests that the window manager not set the input focus to their top-level window.

Maybe worth adding the above (ish) as a comment in the code ?? to help us/others to maintain.

File: include/workspaces.h (hunk 2 above)
Comment (jlindgren90) [reply include/workspaces.h:6]: We presumably don't see a compile error because in each case `wayland-util.h` is already included by some other header before `workspaces.h`.

File: src/xwayland.c (hunk 1 above)
Comment (jlindgren90) [reply src/xwayland.c:40]: Sure, I can do that.

**Comment by jlindgren90** [comment]
> That second commit is hard to follow (for me reading this late in the evening anyway 😄) Is it worth splitting it out into smaller commits doing fewer things?
> 
> EDIT: If possible - it would be easier for my brain to cope if the 'no functional change' (just pure refactoring) commits were separate.

Both these commit are functional changes (I think) -- I tried to put all the pure refactoring in #1109.
Renaming `topmost_mapped` to `topmost_focusable` (or just `topmost`) was meant to reflect the functional change, but I can move the renaming to its own commit if you prefer.

File: include/workspaces.h (hunk 2 above)
Comment (johanmalm) [reply include/workspaces.h:6]: Sorry. Forward declaring is only enough for pointers. Ignore me :blush:
As I said... long day.

**Comment by jlindgren90** [comment]
Rebased again, diff should be smaller now that #1111 is merged.

**Comment by johanmalm** [comment]
LGTM. 
Looks like a merge conflict - guess it's with #1112
